src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp

```diff
@@ -467,6 +469,76 @@ void four_index_constraint(
     }
   }
 }
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+Scalar<DataType> constraint_energy(
+    const tnsr::a<DataType, SpatialDim, Frame>& gauge_constraint,
+    const tnsr::a<DataType, SpatialDim, Frame>& f_constraint,
+    const tnsr::ia<DataType, SpatialDim, Frame>& two_index_constraint,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& three_index_constraint,
+    const tnsr::ijaa<DataType, SpatialDim, Frame>& four_index_constraint,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    double gauge_constraint_multiplier, double two_index_constraint_multiplier,
+    double three_index_constraint_multiplier,
+    double four_index_constraint_multiplier) noexcept {
+  auto energy = make_with_value<Scalar<DataType>>(gauge_constraint, 0.0);
+  constraint_energy<SpatialDim, Frame, DataType>(
+      &energy, gauge_constraint, f_constraint, two_index_constraint,
+      three_index_constraint, four_index_constraint, inverse_spatial_metric,
+      gauge_constraint_multiplier, two_index_constraint_multiplier,
+      three_index_constraint_multiplier, four_index_constraint_multiplier);
+  return energy;
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+void constraint_energy(
+    gsl::not_null<Scalar<DataType>*> energy,
+    const tnsr::a<DataType, SpatialDim, Frame>& gauge_constraint,
+    const tnsr::a<DataType, SpatialDim, Frame>& f_constraint,
+    const tnsr::ia<DataType, SpatialDim, Frame>& two_index_constraint,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& three_index_constraint,
+    const tnsr::ijaa<DataType, SpatialDim, Frame>& four_index_constraint,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    double gauge_constraint_multiplier, double two_index_constraint_multiplier,
+    double three_index_constraint_multiplier,
+    double four_index_constraint_multiplier) noexcept {
+  if (get_size(get(*energy)) != get_size(get<0>(f_constraint))) {
+    *energy = Scalar<DataType>{get<0>(f_constraint)};
+  }
+  get(*energy) = 0.0;
```

Comment:
just don't use `+=`

Reply:
I don't understand this comment. How can I do the summations efficiently without +=?

Reply:
Seems cumbersome, since I need lots of terms with a=0. Is there a big expected penalty to my way of setting to zero?

Reply:
How is it cumbersome? I just gave you the code you need.

Reply:
You're doing work that's not necessary every time. I'm really not seeing how the below code is so much more cumbersome to read.

```cpp
if (get_size(get(*energy)) != get_size(get<0>(f_constraint))) {
  *energy = Scalar<DataType>(get_size(get<0>(f_constraint)));
}

get(*energy) = gauge_constraint_multiplier * square(gauge_constraint.get(0)) +
               two_index_constraint_multiplier * square(f_constraint.get(0));
for (size_t a = 1; a < SpatialDim + 1; ++a) {
  get(*energy) +=
      gauge_constraint_multiplier * square(gauge_constraint.get(a)) +
      two_index_constraint_multiplier * square(f_constraint.get(a));
}

for (size_t a = 0; a < SpatialDim + 1; ++a) {
  for (size_t i = 0; i < SpatialDim; ++i) {
    for (size_t j = 0; j < SpatialDim; ++j) {
      get(*energy) +=
          two_index_constraint_multiplier * two_index_constraint.get(i, a) *
          two_index_constraint.get(j, a) * inverse_spatial_metric.get(i, j);

      for (size_t b = 0; b < SpatialDim + 1; ++b) {
        get(*energy) += three_index_constraint_multiplier *
                            three_index_constraint.get(i, a, b) *
                            three_index_constraint.get(j, a, b) *
                            inverse_spatial_metric.get(i, j) +
                        2.0 * four_index_constraint_multiplier *
                            get(spatial_metric_determinant) *
                            four_index_constraint.get(i, a, b) *
                            four_index_constraint.get(j, a, b) *
                            inverse_spatial_metric.get(i, j);
      }
    }
  }
}
```

Reply:
I feel that splitting the a loop as you do makes it more complicated than keeping the a loop together, but since it sounds like you're saying the extra steps to enable one loop have a likely performance penalty, I guess I don't feel strongly enough about avoiding splitting the loop to insist on it. My latest mixup commit puts this the way you suggested.